File: context-store/src/adapters/redis.ts
import type { ContextStore, SetOptions } from "../types.js";
import { withNamespace, stripNamespace, withTimeout } from "../config.js";

interface Options {
  url: string;
  namespace: string;
  probeTimeoutMs: number;
}

interface RedisLike {
  connect(): Promise<unknown>;
  quit(): Promise<unknown>;
  get(key: string): Promise<string | null>;
  set(key: string, value: string, opts?: { PX?: number }): Promise<unknown>;
  del(key: string): Promise<number>;
  scanIterator(opts: { MATCH: string }): AsyncIterable<string>;
  ping(): Promise<string>;
  on(event: string, listener: (err: unknown) => void): unknown;
}

interface RedisModule {
  createClient(opts: { url: string }): RedisLike;
}

async function loadRedis(): Promise<RedisModule> {
  try {
    return (await import("redis")) as unknown as RedisModule;
  } catch {
    throw new Error(
      "redis backend requested but the 'redis' package is not installed. " +
        "Run `npm install redis` or unset CCC_REDIS_URL.",
    );
  }
}

export async function openRedis({
  url,
  namespace,
  probeTimeoutMs,
}: Options): Promise<ContextStore> {
  const mod = await loadRedis();
  const client = mod.createClient({ url });
  client.on("error", () => {
    /* surfaced via probe / ops */
  });
  await withTimeout(client.connect(), probeTimeoutMs, "redis connect");
  await withTimeout(client.ping(), probeTimeoutMs, "redis ping");

  return {
    kind: "redis",
    async get(key) {
      return client.get(withNamespace(namespace, key));
    },
    async set(key, value, opts: SetOptions = {}) {
      const args = opts.ttlMs ? { PX: opts.ttlMs } : undefined;
      await client.set(withNamespace(namespace, key), value, args);
    },
    async delete(key) {
      const n = await client.del(withNamespace(namespace, key));
      return n > 0;
    },
    async list(prefix) {
      const match = withNamespace(namespace, `${prefix ?? ""}*`);
      const keys: string[] = [];
      for await (const k of client.scanIterator({ MATCH: match })) {
        keys.push(stripNamespace(namespace, k));
      }
      return keys;
    },
    async close() {
      await client.quit();
    },
  };
}

File: src/ccc/store/adapters/redis.py
"""Redis adapter (lazy import of the `redis` package)."""

from __future__ import annotations

from ccc.store.config import strip_namespace, with_namespace
from ccc.store.types import ContextStore


def _import_redis():
    try:
        import redis  # type: ignore
    except ImportError as exc:
        raise RuntimeError(
            "redis backend requested but the 'redis' package is not installed. "
            "Run `pip install redis` or unset CCC_REDIS_URL."
        ) from exc
    return redis


class RedisAdapter(ContextStore):
    kind = "redis"

    def __init__(self, *, url: str, namespace: str, probe_timeout_ms: int) -> None:
        redis = _import_redis()
        timeout_s = max(probe_timeout_ms / 1000, 0.1)
        self._client = redis.Redis.from_url(
            url,
            socket_connect_timeout=timeout_s,
            socket_timeout=timeout_s,
            decode_responses=True,
        )
        self._client.ping()
        self._namespace = namespace

    def get(self, key: str) -> str | None:
        return self._client.get(with_namespace(self._namespace, key))

    def set(self, key: str, value: str, *, ttl_ms: int | None = None) -> None:
        self._client.set(
            with_namespace(self._namespace, key),
            value,
            px=ttl_ms,
        )

    def delete(self, key: str) -> bool:
        return self._client.delete(with_namespace(self._namespace, key)) > 0

    def list(self, prefix: str | None = None) -> list[str]:
        match = with_namespace(self._namespace, f"{prefix or ''}*")
        return [
            strip_namespace(self._namespace, k)
            for k in self._client.scan_iter(match=match)
        ]

    def close(self) -> None:
        try:
            self._client.close()
        except Exception:
            pass


def open_redis(*, url: str, namespace: str, probe_timeout_ms: int) -> RedisAdapter:
    return RedisAdapter(url=url, namespace=namespace, probe_timeout_ms=probe_timeout_ms)
